packages/common/src/GRPCClient.ts

```diff
@@ -1047,6 +1071,199 @@ export default class ConcordiumNodeClient {
+            const bi: v1.FinalizedBlockInfo = { hash, height };
+            height += 1n;
+
+            return bi;
+        };
+
+        const nextNew = async (): Promise<
+            v1.FinalizedBlockInfo | undefined
+        > => {
+            // At this point, we've found all blocks already finalized on chain. Start streaming new blocks.
+            for await (const block of newBlocks) {
+                if (block.height < height) {
+                    // Skip blocks already found.
+                    continue;
+                }
+
+                return block;
+            }
+        };
+
+        const next = async (): Promise<
+            IteratorResult<v1.FinalizedBlockInfo>
+        > => {
+            if (abortSignal?.aborted) {
+                return endSignal;
+            }
+
+            if (finHeight === undefined) {
+                finHeight = await this.getConsensusHeight();
+            }
+
+            let bi: v1.FinalizedBlockInfo | undefined;
+            if (searchKnown) {
+                bi = (await nextKnown()) ?? (await nextNew());
+            } else {
+                bi = await nextNew();
+            }
+
+            if (bi === undefined) {
+                return endSignal;
+            }
+
+            return {
+                done: false,
+                value: bi,
+            };
+        };
+
+        return {
+            [Symbol.asyncIterator]: () => ({ next }),
+        };
+    }
+
+    /**
+     * Find a block with lowest possible height where the predicate holds.
+     *
+     * @template R
+     * @param {(bi: v1.FinalizedBlockInfo) => Promise<R | undefined>} predicate - A predicate function resolving with value of type {@link R} if the predicate holds, and undefined if not.
+     * The precondition for this method is that the function is monotone, i.e., if block at height `h` satisfies the test then also a block at height `h+1` does.
+     * If this precondition does not hold then the return value from this method is unspecified.
+     * @param {bigint} [from=0n] - An optional lower bound of the range of blocks to search. Defaults to 0n.
+     * @param {bigint} [to] - An optional upper bound of the range of blocks to search. Defaults to latest finalized block.
+     *
+     * @returns {Promise<R | undefined>} The value returned from `predicate` at the lowest block (in terms of height) where the predicate holds.
+     */
+    async findEarliestFinalized<R>(
+        predicate: (bi: v1.FinalizedBlockInfo) => Promise<R | undefined>,
+        from: v1.AbsoluteBlocksAtHeightRequest = 0n,
+        to?: v1.AbsoluteBlocksAtHeightRequest
+    ): Promise<R | undefined> {
+        let lower = from;
+        let upper = to ?? (await this.getConsensusHeight());
+
+        if (lower > upper) {
+            throw new Error(
+                'Please specify a "to" value greater than the specified "from" value'
+            );
+        }
+
+        let result: R | undefined;
+        while (lower < upper) {
```

Comment:
This won't work in the case where you the earliest finalized block is the last finalized block, this should work:
```ts
        let result: R | undefined;
        while (lower <= upper) {
            const mid = lower + (upper - lower) / 2n;
            const [hash] = await this.getBlocksAtHeight(mid);
            const res = await predicate({ hash, height: mid });

            if (upper === mid) {
                result = res;
                break;
            } else if (res !== undefined) {
                result = res;
                upper = mid;
            } else {
                lower = mid + 1n;
            }
        }
```

Maybe it can be done more succinctly than mine, but this should at least work :thinking: